File: runners/dotnet/StepDefinitions/AssertionSteps.cs
using TechTalk.SpecFlow;
using OpenFga.Sdk.Model;
using OpenFga.Sdk.Conformance.Support;
using FluentAssertions;
using System.Text.Json;

namespace OpenFga.Sdk.Conformance.StepDefinitions
{
    [Binding]
    public sealed class AssertionSteps
    {
        private readonly TestContext _testContext;

        public AssertionSteps(TestContext testContext)
        {
            _testContext = testContext;
        }

        // Response assertions
        [Then(@"the response should be successful")]
        public void ThenTheResponseShouldBeSuccessful()
        {
            _testContext.AssertResponseSuccess();
        }

        [Then(@"the response should fail")]
        public void ThenTheResponseShouldFail()
        {
            _testContext.AssertResponseFailure();
        }

        [Then(@"the response should have status code (\d+)")]
        public void ThenTheResponseShouldHaveStatusCode(int statusCode)
        {
            var actualStatusCode = _testContext.GetStatusCode();
            if (!actualStatusCode.HasValue)
            {
                throw new AssertionException("No response or error to check status code");
            }
            actualStatusCode.Value.Should().Be(statusCode);
        }

        [Then(@"the response should complete within (\d+) seconds")]
        public void ThenTheResponseShouldCompleteWithin(int seconds)
        {
            // This would require timing implementation in the test context
            // For now, just pass if we have a response
            if (_testContext.LastResponse == null && _testContext.LastError == null)
            {
                throw new AssertionException("No response received");
            }
        }

        [Then(@"the result should be ""([^""]*)""")]
        public void ThenTheResultShouldBe(string expected)
        {
            _testContext.AssertResponseSuccess();

            // Parse expected result (e.g., "allowed: true")
            var parts = expected.Split(": ");
            parts.Should().HaveCount(2, "Expected format: 'field: value'");

            var field = parts[0];
            var value = parts[1];

            if (field == "allowed")
            {
                if (_testContext.LastResponse is CheckResponse checkResponse)
                {
                    var expectedBool = bool.Parse(value);
                    checkResponse.Allowed.Should().Be(expectedBool);
                }
                else
                {
                    throw new AssertionException("Response is not a CheckResponse");
                }
            }
            else
            {
                throw new AssertionException($"Unsupported result field: {field}");
            }
        }

        [Then(@"the response should contain ""([^""]*)""")]
        public void ThenTheResponseShouldContain(string content)
        {
            _testContext.AssertResponseSuccess();
            var responseJson = JsonSerializer.Serialize(_testContext.LastResponse);
            responseJson.Should().Contain(content);
        }

        [Then(@"the response should not contain ""([^""]*)""")]
        public void ThenTheResponseShouldNotContain(string content)
        {
            _testContext.AssertResponseSuccess();
            var responseJson = JsonSerializer.Serialize(_testContext.LastResponse);
            responseJson.Should().NotContain(content);
        }

        [Then(@"the response should have field ""([^""]*)"" with value ""([^""]*)""")]
        public void ThenTheResponseShouldHaveFieldWithValue(string field, string value)
        {
            _testContext.AssertResponseSuccess();

            // Use reflection or JSON path to navigate nested fields
            var responseJson = JsonSerializer.Serialize(_testContext.LastResponse);
            var jsonDoc = JsonDocument.Parse(responseJson);

            var fieldPath = field.Split('.');
            var current = jsonDoc.RootElement;

            foreach (var part in fieldPath)
            {
                current.TryGetProperty(part, out current).Should().BeTrue($"Field '{part}' not found");
            }

            current.GetString().Should().Be(value);
        }

        // Error assertions
        [Then(@"the response should fail with a validation error")]
        public void ThenTheResponseShouldFailWithValidationError()
        {
            _testContext.AssertResponseFailure();

            if (_testContext.LastError != null)
            {
                _testContext.LastError.Message.Should().ContainAny("validation", "invalid", "bad request");
            }
        }

        [Then(@"the response should fail with an authentication error")]
        public void ThenTheResponseShouldFailWithAuthenticationError()
        {
            _testContext.AssertResponseFailure();
            
            // Check for 401 status or authentication-related error
            if (_testContext.LastError is HttpRequestException)
            {
                _testContext.LastError.Message.Should().ContainAny("401", "unauthorized", "authentication");
            }
        }

        [Then(@"the response should fail with an authorization error")]
        public void ThenTheResponseShouldFailWithAuthorizationError()
        {
            _testContext.AssertResponseFailure();
            
            // Check for 403 status or authorization-related error
            if (_testContext.LastError is HttpRequestException)
            {
                _testContext.LastError.Message.Should().ContainAny("403", "forbidden", "authorization");
            }
        }

        [Then(@"the error code should be ""([^""]*)""")]
        public void ThenTheErrorCodeShouldBe(string expectedCode)
        {
            _testContext.AssertResponseFailure();

            // Extract error code from exception or error response
            // Implementation depends on how OpenFGA .NET SDK structures errors
            if (_testContext.LastError != null)
            {
                _testContext.LastError.Message.Should().Contain(expectedCode);
            }
        }

        [Then(@"the error message should contain ""([^""]*)""")]
        public void ThenTheErrorMessageShouldContain(string expectedMessage)
        {
            _testContext.AssertResponseFailure();

            if (_testContext.LastError != null)
            {
                _testContext.LastError.Message.Should().Contain(expectedMessage);
            }
        }

        // Collection assertions
        [Then(@"the response should contain (\d+) items")]
        public void ThenTheResponseShouldContainItems(int count)
        {
            _testContext.AssertResponseContainsItems(count);
        }

        [Then(@"the response should contain at least (\d+) item")]
        [Then(@"the response should contain at least (\d+) items")]
        public void ThenTheResponseShouldContainAtLeastItems(int count)
        {
            _testContext.AssertResponseSuccess();

            if (_testContext.LastResponse is ReadResponse readResponse)
            {
                var actualCount = readResponse.Tuples?.Count ?? 0;
                actualCount.Should().BeGreaterOrEqualTo(count);
            }
            else
            {
                throw new AssertionException("Response is not a ReadResponse");
            }
        }

        [Then(@"the response should contain at most (\d+) items")]
        public void ThenTheResponseShouldContainAtMostItems(int count)
        {
            _testContext.AssertResponseSuccess();

            if (_testContext.LastResponse is ReadResponse readResponse)
            {
                var actualCount = readResponse.Tuples?.Count ?? 0;
                actualCount.Should().BeLessOrEqualTo(count);
            }
            else
            {
                throw new AssertionException("Response is not a ReadResponse");
            }
        }

        [Then(@"the response should contain exactly (\d+) item")]
        [Then(@"the response should contain exactly (\d+) items")]
        public void ThenTheResponseShouldContainExactlyItems(int count)
        {
            _testContext.AssertResponseContainsItems(count);
        }

        [Then(@"the response should be empty")]
        public void ThenTheResponseShouldBeEmpty()
        {
            _testContext.AssertResponseContainsItems(0);
        }

        [Then(@"the tuples should include:")]
        public void ThenTheTuplesShouldInclude(Table table)
        {
            var expectedTuples = new List<TupleKey>();

            foreach (var row in table.Rows)
            {
                var values = row.Values.ToArray();
                expectedTuples.Add(new TupleKey
                {
                    User = values[0],
                    Relation = values[1],
                    Object = values[2]
                });
            }

            _testContext.AssertTuplesInclude(expectedTuples);
        }

        // Header verification
        [Then(@"the request should include header ""([^""]*)"" with value ""([^""]*)""")]
        public void ThenTheRequestShouldIncludeHeaderWithValue(string header, string value)
        {
            _testContext.Headers.Should().ContainKey(header);
            _testContext.Headers[header].Should().Be(value);
        }

        [Then(@"the request should include header ""([^""]*)"" matching ""([^""]*)""")]
        public void ThenTheRequestShouldIncludeHeaderMatching(string header, string pattern)
        {
            _testContext.Headers.Should().ContainKey(header);
            var headerValue = _testContext.Headers[header];
            headerValue.Should().MatchRegex(pattern);
        }

        [Then(@"the request should not include header ""([^""]*)""")]
        public void ThenTheRequestShouldNotIncludeHeader(string header)
        {
            _testContext.Headers.Should().NotContainKey(header);
        }

        [Then(@"the response should include header ""([^""]*)""")]
        public void ThenTheResponseShouldIncludeHeader(string header)
        {
            // This would require response header inspection
            // Implementation depends on how the .NET SDK exposes response headers
            if (_testContext.LastResponse != null)
            {
                // Check if response has headers property or similar
                // For now, just pass - would need SDK-specific implementation
            }
        }

        // Context and state management
        [When(@"I save the response as ""([^""]*)""")]
        public void WhenISaveTheResponseAs(string key)
        {
            _testContext.SaveResponse(key);
        }

        [When(@"I use the saved ""([^""]*)"" for comparison")]
        public void WhenIUseTheSavedForComparison(string key)
        {
            if (!_testContext.SavedData.ContainsKey(key))
            {
                throw new InvalidOperationException($"No saved data found for key: {key}");
            }
        }

        [Then(@"the response should match the saved ""([^""]*)""")]
        public void ThenTheResponseShouldMatchTheSaved(string key)
        {
            _testContext.CompareWithSaved(key, true);
        }

        [Then(@"the response should not match the saved ""([^""]*)""")]
        public void ThenTheResponseShouldNotMatchTheSaved(string key)
        {
            _testContext.CompareWithSaved(key, false);
        }
    }
}

File: runners/python/steps/assertion_steps.py
from behave import given, when, then
from assertpy import assert_that
import json
import re


# Response assertions
@then('the response should be successful')
def step_response_successful(context):
    context.test_context.assert_response_success()


@then('the response should fail')
def step_response_fail(context):
    context.test_context.assert_response_failure()


@then('the response should have status code {status_code:d}')
def step_response_status_code(context, status_code):
    actual_status_code = context.test_context.get_status_code()
    if actual_status_code is None:
        raise AssertionError("No response or error to check status code")
    assert_that(actual_status_code).is_equal_to(status_code)


@then('the response should complete within {seconds:d} seconds')
def step_response_complete_within(context, seconds):
    # This would require timing implementation in the test context
    # For now, just pass if we have a response
    if not context.test_context.last_response and not context.test_context.last_error:
        raise AssertionError("No response received")


@then('the result should be "{expected}"')
def step_result_should_be(context, expected):
    context.test_context.assert_response_success()
    
    # Parse expected result (e.g., "allowed: true")
    parts = expected.split(': ')
    assert len(parts) == 2, f"Expected format: 'field: value', got: {expected}"
    
    field, value = parts
    
    if field == 'allowed':
        expected_bool = value.lower() == 'true'
        if hasattr(context.test_context.last_response, 'allowed'):
            assert_that(context.test_context.last_response.allowed).is_equal_to(expected_bool)
        else:
            raise AssertionError("Response does not have 'allowed' field")
    else:
        raise AssertionError(f"Unsupported result field: {field}")


@then('the response should contain "{content}"')
def step_response_contain(context, content):
    context.test_context.assert_response_success()
    response_json = json.dumps(context.test_context.last_response.__dict__, default=str)
    assert_that(response_json).contains(content)


@then('the response should not contain "{content}"')
def step_response_not_contain(context, content):
    context.test_context.assert_response_success()
    response_json = json.dumps(context.test_context.last_response.__dict__, default=str)
    assert_that(response_json).does_not_contain(content)


@then('the response should have field "{field}" with value "{value}"')
def step_response_field_value(context, field, value):
    context.test_context.assert_response_success()
    
    # Navigate nested fields using dot notation
    field_path = field.split('.')
    current = context.test_context.last_response
    
    for part in field_path:
        assert hasattr(current, part), f"Field '{part}' not found"
        current = getattr(current, part)
    
    assert_that(str(current)).is_equal_to(value)


# Error assertions
@then('the response should fail with a validation error')
def step_response_validation_error(context):
    context.test_context.assert_response_failure()
    
    if context.test_context.last_error:
        error_message = str(context.test_context.last_error).lower()
        assert_that(error_message).matches(r'.*(validation|invalid|bad request).*')


@then('the response should fail with an authentication error')
def step_response_authentication_error(context):
    context.test_context.assert_response_failure()
    
    if context.test_context.last_error:
        error_message = str(context.test_context.last_error).lower()
        assert_that(error_message).matches(r'.*(401|unauthorized|authentication).*')


@then('the response should fail with an authorization error')
def step_response_authorization_error(context):
    context.test_context.assert_response_failure()
    
    if context.test_context.last_error:
        error_message = str(context.test_context.last_error).lower()
        assert_that(error_message).matches(r'.*(403|forbidden|authorization).*')


@then('the error code should be "{expected_code}"')
def step_error_code(context, expected_code):
    context.test_context.assert_response_failure()
    
    if context.test_context.last_error:
        error_message = str(context.test_context.last_error)
        assert_that(error_message).contains(expected_code)


@then('the error message should contain "{expected_message}"')
def step_error_message_contain(context, expected_message):
    context.test_context.assert_response_failure()
    
    if context.test_context.last_error:
        error_message = str(context.test_context.last_error)
        assert_that(error_message).contains(expected_message)


# Collection assertions
@then('the response should contain {count:d} items')
def step_response_contain_items(context, count):
    context.test_context.assert_response_contains_items(count)


@then('the response should contain at least {count:d} item')
@then('the response should contain at least {count:d} items')
def step_response_at_least_items(context, count):
    context.test_context.assert_response_success()
    
    if hasattr(context.test_context.last_response, 'tuples'):
        actual_count = len(context.test_context.last_response.tuples or [])
        assert_that(actual_count).is_greater_than_or_equal_to(count)
    else:
        raise AssertionError("Response does not contain tuples")


@then('the response should contain at most {count:d} items')
def step_response_at_most_items(context, count):
    context.test_context.assert_response_success()
    
    if hasattr(context.test_context.last_response, 'tuples'):
        actual_count = len(context.test_context.last_response.tuples or [])
        assert_that(actual_count).is_less_than_or_equal_to(count)
    else:
        raise AssertionError("Response does not contain tuples")


@then('the response should contain exactly {count:d} item')
@then('the response should contain exactly {count:d} items')
def step_response_exactly_items(context, count):
    context.test_context.assert_response_contains_items(count)


@then('the response should be empty')
def step_response_empty(context):
    context.test_context.assert_response_contains_items(0)


@then('the tuples should include')
def step_tuples_include(context):
    expected_tuples = []
    
    for row in context.table:
        values = list(row.cells)
        expected_tuples.append({
            'user': values[0],
            'relation': values[1],
            'object': values[2]
        })
    
    context.test_context.assert_tuples_include(expected_tuples)


# Header verification
@then('the request should include header "{header}" with value "{value}"')
def step_request_header_value(context, header, value):
    assert_that(context.test_context.headers).contains_key(header)
    assert_that(context.test_context.headers[header]).is_equal_to(value)


@then('the request should include header "{header}" matching "{pattern}"')
def step_request_header_pattern(context, header, pattern):
    assert_that(context.test_context.headers).contains_key(header)
    header_value = context.test_context.headers[header]
    assert_that(header_value).matches(pattern)


@then('the request should not include header "{header}"')
def step_request_not_include_header(context, header):
    assert_that(context.test_context.headers).does_not_contain_key(header)


@then('the response should include header "{header}"')
def step_response_include_header(context, header):
    header_value = context.test_context.get_response_header(header)
    assert_that(header_value).is_not_none()


# Context and state management
@when('I save the response as "{key}"')
def step_save_response(context, key):
    context.test_context.save_response(key)


@when('I use the saved "{key}" for comparison')
def step_use_saved_for_comparison(context, key):
    if key not in context.test_context.saved_data:
        raise Exception(f"No saved data found for key: {key}")


@then('the response should match the saved "{key}"')
def step_response_match_saved(context, key):
    context.test_context.compare_with_saved(key, True)


@then('the response should not match the saved "{key}"')
def step_response_not_match_saved(context, key):
    context.test_context.compare_with_saved(key, False)
